presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -69,47 +69,116 @@ class PartitionAndSerializeOperator : public Operator {
     return !input_;
   }
 
+  // When a new input is added to the operator, we need to reset a few
+  // counters/states.
   void addInput(RowVectorPtr input) override {
     input_ = std::move(input);
-  }
-
-  RowVectorPtr getOutput() override {
-    if (!input_) {
-      return nullptr;
+    numInput = input_->size();
+    if (replicateNullsAndAny_) {
+      collectNullRows();
     }
-    collectNullRows();
+    thisNullCount = nullRows_.countSelected();
+    totalNullCount += thisNullCount;
+    currentState = ProcessingState::kProcessingInputElements;
 
-    const vector_size_t maxOutputSize = calculateMaxOutputSize();
+    velox::row::UnsafeRowFast unsafeRow(reorderInputsIfNeeded());
+    calculateSerializedRowSize(unsafeRow);
 
-    // TODO Reuse output vector.
-    auto output =
-        BaseVector::create<RowVector>(outputType_, maxOutputSize, pool());
-    auto partitionVector = output->childAt(0)->asFlatVector<int32_t>();
-    partitionVector->resize(maxOutputSize);
-    auto dataVector = output->childAt(1)->asFlatVector<StringView>();
-    dataVector->resize(maxOutputSize);
+    outputVector = createOutputRowVector(numInput);
+    auto partitionVector = outputVector->childAt(0)->asFlatVector<int32_t>();
+    auto dataVector = outputVector->childAt(1)->asFlatVector<StringView>();
 
     computePartitions(*partitionVector);
-    serializeRows(*dataVector);
-
-    if (replicateNullsAndAny_ && numPartitions_ > 1) {
-      auto numOutputRows =
-          replicateRowsWithNullPartitionKeys(*dataVector, *partitionVector);
-      // If numOutputRows equal to the input row count, then
-      // 'replicateRowsWithNullPartitionKeys' did not insert any row into the
-      // output vector, thus we need to call into
-      // 'replicateAnyToEmptyPartitions'
-      if (numOutputRows == input_->size()) {
-        numOutputRows = replicateAnyToEmptyPartitions(
-            *partitionVector, *dataVector, numOutputRows);
+    serializeRows(unsafeRow, *dataVector);
+    numOutputRows = numInput;
+    if (replicateNullsAndAny_) {
+      nonEmptyPartitions.insert(
+          partitionVector->rawValues(),
+          partitionVector->rawValues() + numInput);
+      captureReplicateAnyElement(*dataVector);
+      auto averageRowSize = cumulativeRowSizes_[cumulativeRowSizes_.back()] /
+          cumulativeRowSizes_.size();
+      if (thisNullCount > 0) {
+        desiredOutputSize = outputBatchRows(averageRowSize);
       }
-      partitionVector->resize(numOutputRows);
-      dataVector->resize(numOutputRows);
-      output->resize(numOutputRows);
     }
+  }
 
-    input_.reset();
-    return output;
+  /// PartitionAndSerialize is a state machine, the state is maintained in
+  /// 'currentState'. 'getOutput' will generate the outputVector based on
+  /// 'currentState'. Overall, there are 5 states:
+  ///
+  /// kUnknown) This is an unknown state and can only be valid during
```

Comment:
Perhaps, move these comments to ProcessingState. This way there would be no duplication or divergence.